### ISSUE TITLE
Support drawing the tree as ultrametric

### ISSUE THREAD

**Issue by fedarko** [biocore/empress#279]
Suggestion from @esayyari during this morning's meeting.

This would be similar to #271, but -- rather than each branch's length being equal -- the distance from the root to each tip is equal across all tips.

This would probably be another good thing to add in after JS layouts are supported (#213).

**Edit:** [here's](http://phylobotanist.blogspot.com/2018/04/time-calibrated-or-at-least-ultrametric.html) a link to a really nice blog post describing a bunch of different ways to make trees ultrametric. This is probably going to require exposing a few different options to the user in the interface, rather than a single "Ultrametric or not ultrametric?" checkbox.

**Comment by wasade** [comment]
Unit normalization may also make sense (e.g., div all branches by max tip->root distance)

**Comment by gwarmstrong** [comment]
It looks like branch lengths are set in the layout functions in layouts-util.js in [rectangularLayout](https://github.com/biocore/empress/blob/13438f9646b43ed79651624b827b92ca25ae9e4e/empress/support_files/js/layouts-util.js#L133), [circularLayout](https://github.com/biocore/empress/blob/13438f9646b43ed79651624b827b92ca25ae9e4e/empress/support_files/js/layouts-util.js#L365), and [unrootedLayout](https://github.com/biocore/empress/blob/13438f9646b43ed79651624b827b92ca25ae9e4e/empress/support_files/js/layouts-util.js#L543).

All of those lines are the same and could probably be moved to a function call that determines node lengths, i.e., 
```js
function getNodeLengths(ignoreLength, tree, position, precomputedUltrametricLength=null) {
  var nodeLen;
  if (precomputedUltrametricLength === null) {
     nodeLen = ignoreLength ? 1 : tree.length(position);
  } else {
    nodeLen = precomputedUltrametricLength[position];
  }
  return nodeLen
}
```
Where `precomputedUltrametricLength` is computed before the call to this function, depending on the tree should be drawn as ultrametric. Passed as `null` (or not passed) if the tree should not be ultrametric.

**Comment by gibsramen** [comment]
> Edit: here's a link to a really nice blog post describing a bunch of different ways to make trees ultrametric. This is probably going to require exposing a few different options to the user in the interface, rather than a single "Ultrametric or not ultrametric?" checkbox.

I think this may be infeasible @fedarko. Looking at the methods outlined in the blogpost they use penalized likelihood approaches that I think will scale pretty poorly. Talked to @gwarmstrong about this earlier and he devised an algorithm to "effectively stretch each intermediate branch so that its longest root to tip distance is now at the depth of the longest branch."

```
init max_node_to_tip_in_children = 0
for node in PostOrderTraversal(tree):
    node.max_node_to_tip_in_children = max(d(child, node) + child.max_node_to_tip_in_children for child in node.children) max_node_to_tip_distance = root.max_node_to_tip_in_children
root.remaining_distance = 0
for node in PreOrderTraversal(tree):
    node.length += max_node_to_tip_distance - node.max_node_to_tip_in_children - node.depth
```

Implementing this in R right now for ease of visualization and if it works out well we'll port the code to JS.

**Comment by fedarko** [comment]
That's fair -- I guess we won't reliably have bootstrap values (or other relevant likelihood values? I don't really know much about phylogenetics :P) available to us in the first place, and even if we would I agree that some of these methods might be pretty slow to do in the browser.

I'm not super sure how the pseudocode works (looks like the formatting on line 3 might have gotten messed up a bit)?, but as long as it produces _an_ ultrametric tree I think it'll probably be ok. Once we have just one method for doing this in, if people have further requests for algorithms they'd like to see then it shouldn't be too much work to add support for multiple "make ultrametric" algorithms.

Thanks for working on this!

**Comment by gwarmstrong** [comment]
Sweet. Here are some screenshots of the current prototype:
![Screen Shot 2020-11-09 at 4 26 57 PM](https://user-images.githubusercontent.com/19470970/98613949-40188480-22ac-11eb-896b-c4ee2988a893.png)
![Screen Shot 2020-11-09 at 4 32 36 PM](https://user-images.githubusercontent.com/19470970/98613959-43137500-22ac-11eb-90a4-f4e789904141.png)

**Comment by fedarko** [comment]
THAT LOOKS SO COOL AHHH

Thanks so much! It'll be awesome to have this functionality available :)

**Comment by fedarko** [comment]
Closed in #444!
